test/DynamoCoreWpfTests/DynamoTestUIBase.cs

```diff
@@ -24,6 +24,82 @@
 
 namespace DynamoCoreWpfTests
 {
+    internal class TestDiagnostics
+    {
+        internal int DispatcherOpsCounter = 0;
+        // Use this flag to skip trying to execute all the dispatched operations during the test lifetime.
+        // This flag should only be used very sparingly
+        internal bool SkipDispatcherFlush = false;
+
+        private void Hooks_OperationPosted(object sender, DispatcherHookEventArgs e)
+        {
+            e.Operation.Task.ContinueWith((t) => {
+                Interlocked.Decrement(ref DispatcherOpsCounter);
+            }, TaskScheduler.Default);
+            Interlocked.Increment(ref DispatcherOpsCounter);
+        }
+
+        private void PrettyPrint(object obj)
+        {
+            Type type = obj.GetType();
+            PropertyInfo[] properties = type.GetProperties();
+
+            Console.WriteLine("{");
+            foreach (var property in properties)
+            {
+                try
+                {
+                    Console.WriteLine($"  {property.Name}: {property.GetValue(obj)}");
+                }
+                catch (Exception e)
+                {
+                    Console.WriteLine($"  {property.Name}: {e.Message}");
+                }
+            }
+            Console.WriteLine("}");
+        }
+
+        internal void SetupStartupDiagnostics()
+        {
+            System.Console.WriteLine($"PID {Process.GetCurrentProcess().Id} Start test: {TestContext.CurrentContext.Test.Name}");
+            TestUtilities.WebView2Tag = TestContext.CurrentContext.Test.Name;
+
+            SynchronizationContext.SetSynchronizationContext(new DispatcherSynchronizationContext());
+
+            Dispatcher.CurrentDispatcher.Hooks.OperationPosted += Hooks_OperationPosted;
+        }
+
+        internal void SetupBeforeCleanupDiagnostics()
+        {
+            Dispatcher.CurrentDispatcher.Hooks.OperationPosted -= Hooks_OperationPosted;
+            if (!SkipDispatcherFlush)
+            {
+                DispatcherUtil.DoEventsLoop(() => DispatcherOpsCounter == 0);
+            }
+        }
+
+        internal void SetupAfterCleanupDiagnostics()
+        {
+            TestUtilities.WebView2Tag = string.Empty;
+            using (var currentProc = Process.GetCurrentProcess())
+            {
+                int id = currentProc.Id;
+                var name = TestContext.CurrentContext.Test.Name;
+                System.Console.WriteLine($"PID {id} Finished test: {name} with DispatcherOpsCounter = {DispatcherOpsCounter}");
+                System.Console.WriteLine($"PID {id} Finished test: {name} with WorkingSet = {currentProc.WorkingSet64}");
+                System.Console.WriteLine($"PID {id} Finished test: {name} with PrivateBytes = {currentProc.PrivateMemorySize64}");
+                System.Console.Write($"PID {id} Finished test: {name} with GC Memory Info: ");
+                PrettyPrint(GC.GetGCMemoryInfo());
+            }
```

Comment:
Are these metrics expected to be nearly the same as before the tests started?

Reply:
In a perfect world, with no memory leaks, yes